api/src/task/notifications/base_notification.py

```diff
@@ -0,0 +1,98 @@
+from src.task.notifications import constants
+from src.task.notifications.config import EmailNotificationConfig
+from src.task.notifications.constants import Metrics, UserEmailNotification
+from src.task.task import Task
+
+logger = logging.getLogger(__name__)
+
+
+class BaseNotificationTask(Task):
+    Metrics = constants.Metrics
+
+    def __init__(
+        self,
+        db_session: db.Session,
+        notification_config: EmailNotificationConfig | None = None,
+    ):
+        super().__init__(
+            db_session,
+        )
+
+        if notification_config is None:
+            notification_config = EmailNotificationConfig()
+        self.notification_config = notification_config
+
+    @abstractmethod
+    def collect_email_notifications(
+        self,
+    ) -> list[UserEmailNotification]:
+        """Collect email notifications for users"""
+        pass
+
+    @abstractmethod
+    def post_notifications_process(self, notifications: list[UserEmailNotification]) -> None:
+        """Fetch collected notifications and prepare email data."""
+        pass
+
+    def send_notifications(self, notifications: list[UserEmailNotification]) -> None:
+        """Send collected notifications to users"""
+
+        for user_notification in notifications:
+            logger.info(
+                "Sending notification to user",
+                extra={
+                    "user_id": user_notification.user_id,
+                },
+            )
+            notification_log = UserNotificationLog(
+                user_notification_log_id=uuid.uuid4(),
+                user_id=user_notification.user_id,
+                notification_reason=user_notification.notification_reason,
+                notification_sent=False,  # Default to False, update on success
+            )
+            self.db_session.add(notification_log)
+            try:
+                send_pinpoint_email_raw(
+                    to_address=user_notification.user_email,
+                    subject=user_notification.subject,
+                    message=user_notification.content,
+                    app_id=self.notification_config.app_id,
+                )
+                logger.info(
+                    "Successfully sent notification to user",
+                    extra={
+                        "user_id": user_notification.user_id,
+                        "notification_reason": user_notification.notification_reason,
+                        "notification_log_id": notification_log.user_notification_log_id,
+                    },
+                )
+                notification_log.notification_sent = True
+                user_notification.is_notified = True
+
+                self.increment(Metrics.USERS_NOTIFIED)
+
+            except Exception as e:
+                # Notification log will be updated in the finally block
+                logger.exception(
+                    f"Failed to send notification email: {str(e)}",
```

Comment:
`logger.exception` already puts the error on the extra object, don't add it twice.
```suggestion
                    "Failed to send notification email",
```